[APM][.NET] Add B3 documentation (#13727)
DataDog/documentation: content/en/tracing/setup_overview/custom_instrumentation/dotnet.md

    }
}
```

> ### Headers extraction and injection

Reviewer: I'm not sure this information belongs here in the custom instrumentation page. I think once we add the recently-discussed API for injecting/extracting headers, we should include code snippets of that here. And then we can link over to the "setup" page for the reader to learn more about configuring the headers style

Author: Thanks. I put it here as it was there in Java and Go as well. 
But indeed, next week, I'll push the new API and document it here. And I can only link this part. I'll see.

Author: Fixed in 42303f0b25c378a5d52566507547057fba1bb8bf

Datadog APM tracer supports [B3][5] and [W3C][6] headers extraction and injection for distributed tracing.

> Distributed headers injection and extraction is controlled by configuring injection/extraction styles. Currently four styles are supported:

Reviewer 2: Linter warning:
> Avoid temporal words like 'Currently'.
```suggestion
Distributed headers injection and extraction is controlled by configuring injection/extraction styles. The .NET Tracer supports four styles:
```

- Datadog: `Datadog`
- B3: `B3`
- W3C: `TraceParent`
- B3 Single Header: `B3SingleHeader` or `B3 single header`

Injection and extraction styles can be configured using:

- Environment Variable: `DD_PROPAGATION_STYLE_INJECT=Datadog, B3, W3C`
- Environment Variable: `DD_PROPAGATION_STYLE_EXTRACT=Datadog, B3, W3C`

The values of these environment variables are comma separated lists of header styles that are enabled for injection or extraction. By default only `Datadog` injection style is enabled.

If multiple extraction styles are enabled extraction attempt is done on the order those styles are configured and first successful extracted value is used.

## Resource filtering

> Traces can be excluded based on their resource name, to remove synthetic traffic such as health checks from reporting traces to Datadog.  This and other security and fine-tuning configurations can be found on the [Security][7] page.

Reviewer 2: Linter error
> Use only one space between words and sentences (not two).
```suggestion
Traces can be excluded based on their resource name, to remove synthetic traffic such as health checks from reporting traces to Datadog. This and other security and fine-tuning configurations can be found on the [Security][7] page.
```

    }
}
```

### Headers extraction and injection

> Datadog APM tracer supports [B3][5] and [W3C][6] headers extraction and injection for distributed tracing. You can find more details on this in the [setup documentation][7].

Reviewer 3: ```suggestion
The Datadog APM Tracer supports [B3][5] and [W3C][6] headers extraction and injection for distributed tracing. For more information, see the [setup documentation][7].
```

Author: Including the change. But note that this text was a copy paste from the Java and Go do IIRC.

> Headers extraction and injection will be transparent in most cases.
> Though, in some rare cases the span context can be lost and only a custom implementation will be able to fix it.
> 
> To do so, please do the following:

Reviewer 3: ```suggestion
In most cases, the header's extraction and injection are transparent. If your span context gets lost, create a custom implementation.
```

Author: 🤔 , I think I need to explicit more when this can happen.
I think I'll rework it to:

> In most cases, the header's extraction and injection are transparent. Though, there are some known cases where your distribute trace can be disconnected. For instance, when reading messages from a distributed queue, some library will lose the span context. In that case, you can add a custom trace using the following code:

WDYT?

```csharp
var spanContextExtractor = new SpanContextExtractor();
var parentContext = spanContextExtractor.Extract(headers, (headers, key) => GetHeaderValues(headers, key));
var spanCreationSettings = new SpanCreationSettings() { Parent = parentContext }; 
using var scope = Tracer.Instance.StartActive("operation", spanCreationSettings);
```

> Where `GetHeaderValues` will depend on the underlying system you are using. Here are a few examples:

Author: Here, it's not a problem of location. But more of definition. The code above contains a function called `GetHeaderValues`. This method will have to be provided by the user.

So I could rephrase to:

> Where `GetHeaderValues` is a method that you will provide. Its implementation will depend on the structure carrying the `SpanContext`. Here are a few examples:

WDYT?

```csharp
// Confluent.Kafka
IEnumerable<string> GetHeaderValues(Headers headers, string name)
{
    if (headers.TryGetLastBytes(name, out var bytes))
    {
        try
        {
            return new[] { Encoding.UTF8.GetString(bytes) };
        }
        catch (Exception)
        {
            // ignored
        }
    }

    return Enumerable.Empty<string>();
}

// RabbitMQ
IEnumerable<string> GetHeaderValues(IDictionary<string, object> headers, string name)
{
    if (headers.TryGetValue(name, out object value) && value is byte[] bytes)
    {
        return new[] { Encoding.UTF8.GetString(bytes) };
    }

    return Enumerable.Empty<string>();
}
```

## Resource filtering

> Traces can be excluded based on their resource name, to remove synthetic traffic such as health checks from reporting traces to Datadog. This and other security and fine-tuning configurations can be found on the [Security][8] page.

Reviewer 3: ```suggestion
You can exclude traces based on the resource name to remove Synthetics traffic such as health checks. For more information about security and additional configurations, see [Configure the Datadog Agent or Tracer for Data Security][8].
```
